File: packages/aap-swift/Sources/AAPKit/AAPSession.swift
import Foundation

/// A live encrypted session between two AAP agents.
///
/// Created by `AAPAgent.connect()` or `AAPAgent.connectViaRelay()`.
/// Do not create directly — use `AAPAgent` as the entry point.
@MainActor
public final class AAPSession: @unchecked Sendable {

    // MARK: - Properties

    private let localIdentity:    AgentIdentity
    private let privateKeyHex:    String
    private let remoteAddress:    String
    private let registryURL:      URL
    private var relay:            AAPRelay?
    private var sessionKey:       String?
    private var remotePublicKey:  String?
    private var _connected        = false

    private var messageHandlers: [(AAPMessage) -> Void] = []

    // MARK: - Init (package-internal)

    init(
        localIdentity: AgentIdentity,
        privateKeyHex: String,
        remoteAddress: String,
        registryURL:   URL,
        relay:         AAPRelay? = nil
    ) {
        self.localIdentity = localIdentity
        self.privateKeyHex = privateKeyHex
        self.remoteAddress = remoteAddress
        self.registryURL   = registryURL
        self.relay         = relay
    }

    // MARK: - Relay handshake

    /// Perform key exchange over the relay. Call after `AAPAgent.connectViaRelay()`.
    func handshakeViaRelay() async throws {
        guard let relay else { throw AAPError.handshakeFailed("No relay attached") }

        let nonce_a = generateNonce()

        // Send HANDSHAKE_INIT and wait for response
        let initMsg: [String: Any] = [
            "type":       "HANDSHAKE_INIT",
            "from_did":   localIdentity.did,
            "nonce_a":    nonce_a,
            "timestamp":  Int64(Date().timeIntervalSince1970 * 1000),
        ]
        let initData = try JSONSerialization.data(withJSONObject: initMsg)
        relay.send(data: initData)

        // Wait for either HANDSHAKE_INIT (we are guest) or HANDSHAKE_RESPONSE (we are host)
        try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Void, Error>) in
            let timer = Task {
                try await Task.sleep(nanoseconds: 15_000_000_000)
                cont.resume(throwing: AAPError.relayTimeout)
            }
            relay.onReceive { [weak self] data in
                guard let self else { return }
                guard let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                      let type = obj["type"] as? String else { return }

                Task { @MainActor in
                    timer.cancel()
                    if type == "HANDSHAKE_RESPONSE",
                       let derivedKey = obj["session_key"] as? String {
                        self.sessionKey = derivedKey
                        self._connected = true
                        cont.resume()
                    } else if type == "HANDSHAKE_INIT",
                              let peerNonce = obj["nonce_a"] as? String,
                              let peerDID   = obj["from_did"] as? String {
                        // We are guest — reply
                        let nonce_b = generateNonce()
                        let combined = peerNonce + nonce_b
                        self.sessionKey  = combined // simplified — real impl uses HKDF
                        self._connected  = true
                        self.remotePublicKey = peerDID

                        let response: [String: Any] = [
                            "type":        "HANDSHAKE_RESPONSE",
                            "nonce_b":     nonce_b,
                            "session_key": combined,
                        ]
                        if let responseData = try? JSONSerialization.data(withJSONObject: response) {
                            relay.send(data: responseData)
                        }
                        cont.resume()
                    }
                }
            }
        }

        print("✅ [AAPKit] Relay tunnel established (\(remoteAddress))")
    }

    // MARK: - Send

    /// Send a typed action to the connected peer.
    public func send(
        actionType: String,
        payload:    [String: AAPValue] = [:]
    ) async throws {
        guard _connected else { throw AAPError.handshakeFailed("Not connected") }

        let envelope = try buildEnvelope(
            actionType:    actionType,
            payload:       payload,
            fromDID:       localIdentity.did,
            toDID:         remoteAddress,
            privateKeyHex: privateKeyHex
        )

        if let relay {
            let encoder    = JSONEncoder()
            encoder.outputFormatting = .sortedKeys
            let envData    = try encoder.encode(envelope)
            relay.send(data: envData)
        } else {
            // Direct HTTP delivery (non-relay sessions)
            throw AAPError.sendFailed("Direct delivery not yet implemented in Swift SDK — use relay")
        }
    }

    // MARK: - Receive

    /// Register a handler for incoming messages.
    public func onMessage(_ handler: @escaping (AAPMessage) -> Void) {
        messageHandlers.append(handler)
    }

    // MARK: - Lifecycle

    public func disconnect() async {
        relay?.close()
        _connected = false
        sessionKey = nil
    }

    public var isConnected: Bool { _connected }
}

File: packages/aap-android/src/main/kotlin/dev/aap/sdk/AAPSession.kt
package dev.aap.sdk

import dev.aap.sdk.crypto.*
import kotlinx.coroutines.*
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*

/**
 * A live encrypted session between two AAP agents.
 * Created by [AAPAgent.connectViaRelay] — do not construct directly.
 */
class AAPSession internal constructor(
    private val localIdentity: AgentIdentity,
    private val privateKeyHex: String,
    private val remoteAddress: String,
    private val relay:         AAPRelay?    = null
) {
    private var sessionKey: String?  = null
    private var _connected           = false
    private val messageHandlers      = mutableListOf<(AAPMessage) -> Unit>()

    private val json = Json { encodeDefaults = true }

    // ── Relay handshake ────────────────────────────────────────────────────────

    internal suspend fun handshakeViaRelay() {
        val r = relay ?: throw AAPError.HandshakeFailed("No relay attached")

        val nonce_a  = generateNonce()
        val initMsg  = buildJsonObject {
            put("type",      "HANDSHAKE_INIT")
            put("from_did",  localIdentity.did)
            put("nonce_a",   nonce_a)
            put("timestamp", System.currentTimeMillis())
        }
        r.send(initMsg.toString())

        withTimeout(15_000) {
            suspendCancellableCoroutine<Unit> { cont ->
                r.onReceive { data ->
                    val text = data.toString(Charsets.UTF_8)
                    try {
                        val obj  = Json.parseToJsonElement(text).jsonObject
                        val type = obj["type"]?.jsonPrimitive?.content
                        if (type == "HANDSHAKE_RESPONSE") {
                            val nonce_b = obj["nonce_b"]?.jsonPrimitive?.content ?: ""
                            sessionKey  = nonce_a + nonce_b   // simplified; real: HKDF
                            _connected  = true
                            // Re-register for data frames
                            r.onReceive { frame -> dispatchMessage(frame) }
                            cont.resume(Unit) {}
                        } else if (type == "HANDSHAKE_INIT") {
                            // We are the guest
                            val peerNonce = obj["nonce_a"]?.jsonPrimitive?.content ?: ""
                            val nonce_b   = generateNonce()
                            sessionKey    = peerNonce + nonce_b
                            _connected    = true
                            val response  = buildJsonObject {
                                put("type",    "HANDSHAKE_RESPONSE")
                                put("nonce_b", nonce_b)
                            }
                            r.send(response.toString())
                            r.onReceive { frame -> dispatchMessage(frame) }
                            cont.resume(Unit) {}
                        }
                    } catch (_: Exception) { /* ignore non-JSON */ }
                }
            }
        }
    }

    // ── Send ───────────────────────────────────────────────────────────────────

    /**
     * Send a typed action to the connected peer.
     * The payload is wrapped in a three-layer AAP envelope before sending.
     */
    suspend fun send(
        actionType: String,
        payload:    Map<String, JsonElement> = emptyMap()
    ) {
        if (!_connected) throw AAPError.HandshakeFailed("Not connected")

        val envelope = buildEnvelope(
            actionType    = actionType,
            payload       = payload,
            fromDID       = localIdentity.did,
            toDID         = remoteAddress,
            privateKeyHex = privateKeyHex
        )
        val frame = json.encodeToString(envelope)

        relay?.send(frame)
            ?: throw AAPError.SendFailed("Direct delivery not yet implemented — use relay")
    }

    // ── Receive ────────────────────────────────────────────────────────────────

    fun onMessage(handler: (AAPMessage) -> Unit) {
        messageHandlers += handler
    }

    private fun dispatchMessage(data: ByteArray) {
        try {
            val outer = Json.decodeFromString<dev.aap.sdk.crypto.OuterEnvelope>(
                data.toString(Charsets.UTF_8)
            )
            val middle = Json.decodeFromString<dev.aap.sdk.crypto.MiddleEnvelope>(outer.middle_payload)
            val inner  = Json.decodeFromString<dev.aap.sdk.crypto.InnerEnvelope>(middle.inner_payload)
            val msg    = AAPMessage(
                actionType = inner.action_type,
                fromDID    = outer.from_did,
                messageID  = outer.message_id,
                timestamp  = inner.timestamp,
                payload    = inner.payload
            )
            messageHandlers.forEach { it(msg) }
        } catch (_: Exception) { /* malformed frame — ignore */ }
    }

    // ── Lifecycle ──────────────────────────────────────────────────────────────

    fun disconnect() {
        relay?.close()
        _connected = false
        sessionKey = null
    }

    val isConnected: Boolean get() = _connected
}
